### PR TITLE
autocomplete module now uses textcomplete instead of jquery-ui.autocomplete

### PR DESCRIPTION
**Pull request by akhoury**
# ~DO NOT MERGE YET~

~Pending issues with `textcomplete` module~
* ~https://github.com/yuku/textcomplete/issues/165~ solved with https://github.com/NodeBB/NodeBB/commit/8c5c4ba40c530262bb8cfc658478eebe1c459cb1
* ~https://github.com/yuku/textcomplete/issues/166~ temporarily solved with this hack https://github.com/NodeBB/NodeBB/commit/278850a5e4e7a29046d7926b8aa38e16f9dc9eaf pending https://github.com/yuku/textcomplete/issues/166

### CODE REVIEW

File: public/src/utils.js [hunk 1]
```diff
@@ -351,7 +351,7 @@
 			if (!str) {
 				return '';
 			}
-			str = str.replace(utils.trimRegex, '');
+			str = ('' + str).replace(utils.trimRegex, '');
```
Comment (akhoury) [inline public/src/utils.js]: ok, so, the reasoning for this one here is the following:

1) in `Privileges.addUserToPrivilegeTable()` the `member` value sent via the socket request is the integer value `uid`, `member: ui.item.user.uid` 

https://github.com/NodeBB/NodeBB/blob/9b308e4b07f2ded770aefc6ffb1b35e5166642ef/public/src/admin/manage/privileges.js#L144-L150

so ... when `Categories.setPrivilege()` is called, notice the `group.exists(data.member)` is called
https://github.com/NodeBB/NodeBB/blob/9b308e4b07f2ded770aefc6ffb1b35e5166642ef/src/socket.io/admin/categories.js#L51-L58

`Group.exists()` internally uses `utils.slugify()` which expects a string only, when getting an `int` we get an error

https://github.com/NodeBB/NodeBB/blob/9b308e4b07f2ded770aefc6ffb1b35e5166642ef/src/groups/index.js#L186-L194

did this come up before, the way to reproduce this is to visit ACP > Manage > Privileges and add a user to a specific category.

Comment (barisusakli) [reply public/src/utils.js]: It's a bug caused by a recent change I made, I will take a look at that https://github.com/NodeBB/NodeBB/commit/3455c27a8d3fe59ddf5791559ba283ccdeeaed80

Comment (barisusakli) [reply public/src/utils.js]: Fixed in https://github.com/NodeBB/NodeBB/commit/441dd86d29812dc5e5167ac7eb415b5c8468bc4e

**Comment by pitaj** [comment]
> temporarily solved with this hack 278850a pending yuku/textcomplete#166

Will this hack break when textcomplete is fixed? If so, might be worth sending a PR to textcomplete to fix it so we don't have to worry about breakage in the future.

**Comment by akhoury** [comment]
@pitaj i did suggest the change to textcomplete, but they might want to tackle it differently. To answer your question, no it should, maybe we can just for the exact version of textcomplete to `"0.18.0"` instead of `"^0.18.0"`

**Comment by julianlam** [comment]
@akhoury Any chance you can re-do this plugin or fix the conflicts? :smile:

**Comment by barisusakli** [comment]
image
I think textcomplete package is no longer updated and it got broken down into multiple packages.

https://github.com/yuku/old-textcomplete#readme

**Comment by julianlam** [comment]
Lovely, lol

**Comment by julianlam** [comment]
Closing for now, we'll want to eventually use @textcomplete/core, but it might need webpack

**Pull request closed without merging**